src/compose/ports.ts

```diff
@@ -0,0 +1,206 @@
+	 * the output of the docker daemon when probed about a given container)
+	 * and produces a list of PortMap objects, which can then be compared.
+	 *
+	 */
+	public static fromDockerOpts(
+		portBindings: PortBindings,
+	): PortMap[] {
+
+		// Create a list of portBindings, rather than the map (which we can't
+		// order)
+		const portMaps = _.map(portBindings, (hostObj, internalStr) => {
+
+			const match = internalStr.match(DOCKER_OPTS_PORTS_REGEX);
+			if (match == null) {
+				throw new Error(`Could not parse docker port output: ${internalStr}`);
+			}
+			const internal = parseInt(match![1], 10);
+			const external = parseInt(hostObj[0].HostPort, 10);
+
+			const host = hostObj[0].HostIp;
+			const proto = match[2] || 'tcp';
+
+			return new PortMap({
+				internalStart: internal,
+				internalEnd: internal,
+				externalStart: external,
+				externalEnd: external,
+				protocol: proto,
+				host,
+			});
+		});
+
+		return PortMap.normalisePortMaps(portMaps);
+	}
+
+	public static normalisePortMaps(portMaps: PortMap[]): PortMap[] {
+		// Fold any ranges into each other if possible
+		return _(portMaps)
+			.sortBy((p) => p.protocol)
+			.sortBy((p) => p.host)
+			.sortBy((p) => p.internalStart)
+			.reduce((res: PortMap[], p: PortMap) => {
+				const last = _.last(res);
+				if (last == null) {
+					res.push(p);
+				} else {
+					if (last.internalEnd + 1 === p.internalStart &&
+						last.externalEnd + 1 === p.externalStart &&
+						last.protocol === p.protocol &&
+						last.host === p.host
+					) {
+						last.internalEnd += 1;
+						last.externalEnd += 1;
+					} else {
+						res.push(p);
+					}
+				}
+				return res;
+			}, []);
+	}
+
+	private parsePortString(portStr: string): void {
+		const match = portStr.match(PORTS_REGEX);
+		if (match == null) {
+			throw new InvalidPortDefinition(`Could not parse port definition: ${portStr}`);
+		}
+
+		let [
+			,
+			host = '',
+			external,
+			externalEnd,
+			internal,
+			internalEnd,
+			protocol = 'tcp',
+		] = match;
+
+		if (external == null) {
+			external = internal;
+		}
+
+		if (internalEnd == null) {
+			internalEnd = internal;
+		}
+
+		if (externalEnd == null) {
+			if (internal === internalEnd) {
+				// This is a special case to handle a:b
+				externalEnd = external;
+			} else {
+				// and this handles a-b
+				externalEnd = internalEnd;
+			}
+		}
+
+		this.internalStart = parseInt(internal, 10);
+		this.internalEnd = parseInt(internalEnd, 10);
+		this.externalStart = parseInt(external, 10);
+		this.externalEnd = parseInt(externalEnd, 10);
+		this.host = host;
+		this.protocol = protocol;
+
+		// Ensure we have the same range
+		if (this.internalEnd - this.internalStart !== this.externalEnd - this.externalStart) {
+			throw new InvalidPortDefinition(
+				`Range for internal and external ports does not match: ${portStr}`,
+			);
+		}
+	}
+
+	private generatePortRange(start: number, end: number): number[] {
```

Comment:
It might be worth adding a check that `start <= end` and throwing an error if not

Reply:
Done